{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "169fdbe7",
   "metadata": {},
   "source": [
    "## Relative Abundance Stacked Barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "06944495",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "48b2d500",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set taxon level\n",
    "# taxon_level = 'genus'\n",
    "taxon_level = 'species'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "ebccab6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load biom table and metadata\n",
    "biom_path = f\"../tables/3_228413_gOTU_micov75%_{taxon_level}_collapsed.biom\"\n",
    "metadata = pd.read_csv(\"../metadata/marta_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "baecd70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View biom table as convert to tsv\n",
    "biom_table = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_table.to_dataframe()).T\n",
    "df.index = df.index.astype(str)\n",
    "df.index = df.index.str.replace('16207.', '')\n",
    "df = df.loc[metadata['sample_name']]\n",
    "# df = df.rename(columns={\"s__\": \"s__Unknown\"})\n",
    "if 's__' in df.columns:\n",
    "    df = df.drop('s__', axis=1)\n",
    "    \n",
    "# Save df as tsv\n",
    "df.to_csv(f\"../tables/3_228413_gOTU_micov75%_{taxon_level}_collapsed.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "33397fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input/output paths\n",
    "infile = f\"../tables/3_228413_gOTU_micov75%_{taxon_level}_collapsed.tsv\"\n",
    "outfile = f\"../tables/4_228413_gOTU_micov75%_{taxon_level}_collapsed_relab.tsv\"\n",
    "\n",
    "# Load table: rows = samples, columns = taxa\n",
    "df = pd.read_csv(infile, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Convert to relative abundance (per sample)\n",
    "df_rel = df.div(df.sum(axis=1), axis=0)\n",
    "\n",
    "# Load metadata and index by sample_name\n",
    "md2 = metadata.set_index(\"sample_name\")\n",
    "\n",
    "# Update index to:  (sample_number) treatment [microbiome_type]\n",
    "sample_name_to_index = {\n",
    "    sn: f\"({md2.loc[sn, 'sample_number']}) {md2.loc[sn, 'treatment']} [{md2.loc[sn, 'microbiome_type']}]\"\n",
    "    for sn in md2.index\n",
    "}\n",
    "\n",
    "df_rel.index = df_rel.index.map(lambda x: sample_name_to_index[x])\n",
    "\n",
    "# Save as TSV\n",
    "df_rel.to_csv(outfile, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "76293745",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_60616/3861053802.py:5: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed two minor releases later. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.\n",
      "  colors = cm.get_cmap('tab20')(np.linspace(0, 1, len(all_species_global)))\n"
     ]
    }
   ],
   "source": [
    "# Get all unique species across the whole dataframe\n",
    "all_species_global = df_rel.columns.tolist()\n",
    "\n",
    "# Create a consistent color palette for all species\n",
    "colors = cm.get_cmap('tab20')(np.linspace(0, 1, len(all_species_global)))\n",
    "species_colors_global = dict(zip(all_species_global, colors))\n",
    "\n",
    "# Loop through each group\n",
    "for group in md2['plot_group'].unique():\n",
    "    # Get samples in this group\n",
    "    samples_in_group = md2[md2['plot_group'] == group].index\n",
    "    # Map to new index names\n",
    "    new_index_names = [sample_name_to_index[s] for s in samples_in_group if s in sample_name_to_index]\n",
    "    \n",
    "    # Subset df_rel\n",
    "    df_group = df_rel.loc[df_rel.index.isin(new_index_names)]\n",
    "    \n",
    "    # Plot with independent sorting for each bar\n",
    "    plt.figure(figsize=(20, 10))\n",
    "    \n",
    "    # Track which species have been added to legend\n",
    "    legend_added = set()\n",
    "    \n",
    "    # For each sample, sort species by abundance and plot\n",
    "    for i, sample in enumerate(df_group.index):\n",
    "        # Sort species for this sample in descending order (most abundant first)\n",
    "        sample_sorted = df_group.loc[sample].sort_values(ascending=False)\n",
    "        \n",
    "        # Initialize bottom position for stacking\n",
    "        bottom = 0\n",
    "        \n",
    "        # Plot each species as a bar segment\n",
    "        for species, abundance in sample_sorted.items():\n",
    "            if abundance > 0:\n",
    "                # Add to legend only once\n",
    "                label = species if species not in legend_added else \"\"\n",
    "                if label:\n",
    "                    legend_added.add(species)\n",
    "                \n",
    "                plt.bar(i, abundance, bottom=bottom, width=0.8,\n",
    "                        color=species_colors_global[species], label=label)\n",
    "                bottom += abundance\n",
    "    \n",
    "    # Set x-axis labels and positions\n",
    "    plt.xticks(range(len(df_group.index)), df_group.index, rotation=90, ha=\"center\", fontsize=16)\n",
    "    plt.ylabel(\"Relative Abundance\", fontsize=16)\n",
    "    plt.title(f\"Microbiome Composition ({taxon_level}): {group}\", fontsize=22)\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc=\"best\", title=taxon_level, title_fontsize=16, fontsize=14)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if group == 'Em23 + MC mouse stool':\n",
    "        plt.text(\n",
    "            1.2,\n",
    "            -0.8,\n",
    "            \"Data by Yang Chen\",\n",
    "            fontsize=10,\n",
    "            color='grey',\n",
    "            ha='right',\n",
    "            va='bottom',\n",
    "            transform=plt.gca().transAxes\n",
    "        )\n",
    "    elif group == 'Em19 mouse skin':\n",
    "        plt.text(\n",
    "            1.2,\n",
    "            -0.4,\n",
    "            \"Data by Yang Chen\",\n",
    "            fontsize=10,\n",
    "            color='grey',\n",
    "            ha='right',\n",
    "            va='bottom',\n",
    "            transform=plt.gca().transAxes\n",
    "        )\n",
    "    elif group == 'Em19 mouse stool':\n",
    "        plt.text(\n",
    "            1.2,\n",
    "            -0.4,\n",
    "            \"Data by Yang Chen\",\n",
    "            fontsize=10,\n",
    "            color='grey',\n",
    "            ha='right',\n",
    "            va='bottom',\n",
    "            transform=plt.gca().transAxes\n",
    "        )            \n",
    "    else:\n",
    "        plt.text(\n",
    "            1.2,\n",
    "            -0.5,\n",
    "            \"Data by Yang Chen\",\n",
    "            fontsize=10,\n",
    "            color='grey',\n",
    "            ha='right',\n",
    "            va='bottom',\n",
    "            transform=plt.gca().transAxes\n",
    "        )\n",
    "\n",
    "    \n",
    "    # Save figure\n",
    "    plt.savefig(f\"../figures/relative_abundance_sorted_{group}_{taxon_level}.png\", dpi=600, bbox_inches='tight')\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
